### PR TITLE
feat(forms): Add a new step to the import process to solve requirements issues

### PR DESCRIPTION
**Pull request by ccailly**
## Checklist before requesting a review

*Please delete options that are not relevant.*

- [X] I have read the CONTRIBUTING document.
- [X] I have performed a self-review of my code.
- [X] I have added tests that prove my fix is effective or that my feature works.

## Description

- Add a new step when importing a form to reconcile objects not available on the GLPI where the form will be imported.
- Add an action to remove a form from the import list.

image
image
image

### CODE REVIEW

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
The UI looks great but I have some request on the technical side.

The idea (which I should have worded directly in the specifications, my bad) was to use the existing DatabaseMapper object to add the replacements items without having to modify the code of the serializer itself.

Something like that:
```php
        $serializer = new FormSerializer();
        $mapper = new DatabaseMapper(Session::getActiveEntities());

+       $replacements = $request->request->all()["replacements"] ?? [];
+       foreach ($replacements as $replacement) {
+           $mapper->addMappedItem(...);
+       }

        return $this->render("pages/admin/form/import/step4_execute.html.twig", [
            'title'   => __("Import results"),
            'menu'    => ['admin', Form::getType()],
            'results' => $serializer->importFormsFromJson($json, $mapper),
        ]);
```

Unless there are some technical difficulties that I am missing, this solution would avoid needing to modify the serializer methods as we reuse the data mapper parameter that was made for this use case (I hope there are no tricky edge case that would make it impossible).

If this kind of solution can work, it will make the code much simpler :)

**Review by AdrienClairembault — COMMENTED** [review]
Regarding the removal of a form, there are two issues:
- We are searching on the form name
- We are editing the JSON

Searching on the form name may be dangerous as it is not guaranteed to be unique.
Maybe we should include some kind of unique ID for each forms into the export instead ?
Then we can exclude form using this id instead of the name.

I'd also like to avoid editing the JSON if possible.
An hidden input with the ids of the ignored forms would be simpler IMO.

Then the serializer can know that the form is ignored and we can display it on the report if you see what I mean ?

File: src/Glpi/Form/Export/Serializer/FormSerializer.php [hunk 1]
```diff
@@ -97,19 +99,64 @@ public function previewImport(
...
+
+        // Filter the forms to remove the one that matches the given name
+        $export_specification->forms = array_filter(
+            $export_specification->forms,
+            fn($form_spec) => $form_spec->name !== $form_name
+        );
+
+        return $this->serialize($export_specification);
+    }
+
```
Comment (AdrienClairembault) [inline src/Glpi/Form/Export/Serializer/FormSerializer.php:156]: This is not longer used.

```suggestion
```

File: src/Glpi/Form/Export/Serializer/FormSerializer.php [hunk 2]
```diff
@@ -97,19 +99,64 @@ public function previewImport(
...
             } else {
-                $results->addInvalidForm($form_name);
+                $results->addInvalidForm($form_id, $form_name);
             }
         }
 
         return $results;
     }
 
+    public function resolveIssues(
```
Comment (AdrienClairembault) [inline src/Glpi/Form/Export/Serializer/FormSerializer.php:119]: Can you rename this method ?
The current name seems to imply that it solve issues when it seems to be listing issues instead.

File: phpunit/functional/Glpi/Form/Export/FormSerializerTest.php
Comment (AdrienClairembault) [inline phpunit/functional/Glpi/Form/Export/FormSerializerTest.php:1]: Can you add a new test case here to show that the skipped form parameter is taken into account ?

File: src/Glpi/Form/Export/Serializer/FormSerializer.php [hunk 3]
```diff
@@ -82,6 +83,7 @@ public function exportFormsToJson(array $forms): ExportResult
 
     public function previewImport(
         string $json,
+        array $skipped_forms,
```
Comment (AdrienClairembault) [inline src/Glpi/Form/Export/Serializer/FormSerializer.php:86]: I think it make sense that the import can be run without ignoring any form, thus maybe this parameter should be optional (move it to last + add `[]` default value) ?

File: src/Glpi/Form/Export/Serializer/FormSerializer.php [hunk 4]
```diff
@@ -97,19 +99,64 @@ public function previewImport(
...
+            $export_specification->forms,
+            fn($form_spec) => $form_spec->name !== $form_name
+        );
+
+        return $this->serialize($export_specification);
+    }
+
     public function importFormsFromJson(
         string $json,
+        array $skipped_forms,
```
Comment (AdrienClairembault) [inline src/Glpi/Form/Export/Serializer/FormSerializer.php:159]: Same here regarding args order.

File: src/Glpi/Form/Export/Serializer/FormSerializer.php [hunk 5]
```diff
@@ -125,6 +172,11 @@ public function importFormsFromJson(
             $requirements = $form_spec->data_requirements;
             $mapper->mapExistingItemsForRequirements($requirements);
 
+            $form_name = $form_spec->name;
+            if (in_array($form_name, $skipped_forms)) {
+                continue;
+            }
+
```
Comment (AdrienClairembault) [inline src/Glpi/Form/Export/Serializer/FormSerializer.php:179]: I think you forgot to switch to IDs here ;).
(It also mean the test suite is incomplete as it should catch this error).

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]

**Review by AdrienClairembault — APPROVED** [review]

File: templates/pages/admin/form/import/step3_resolve_issues.html.twig [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+{% extends "layout/page_without_tabs.html.twig" %}
+
+{% set container_size = "narrow" %}
+
+{% block content_title %}
+	{{ __('Import forms') }}
+{% endblock content_title %}
+
+{% block content_body %}
+	<form method="POST" action="{{ path('Form/Import/Preview') }}">
```
Comment (cedric-anne) [inline templates/pages/admin/form/import/step3_resolve_issues.html.twig:45]: This file is indented using tabs.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**